applications/CompressiblePotentialFlowApplication/custom_elements/embedded_transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,563 @@
+        }
+    }
+}
+
+template <int TDim, int TNumNodes>
+void EmbeddedTransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::FindUpwindElement(const ProcessInfo& rCurrentProcessInfo)
+{
+    GeometryType upwind_element_boundary=this->GetGeometry();
+    std::vector<size_t> upwind_element_nodes;
+    GlobalPointersVector<Element> upwind_element_candidates;
+    PotentialFlowUtilities::GetNodeNeighborElementCandidates<TDim, TNumNodes>(upwind_element_candidates, upwind_element_boundary);
+
+    SelectUpwindElement(upwind_element_nodes, upwind_element_candidates, rCurrentProcessInfo);
+}
+
+template <int TDim, int TNumNodes>
+void EmbeddedTransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::SelectUpwindElement(
+    std::vector<IndexType>& rUpwindElementNodeIds,
+    GlobalPointersVector<Element>& rUpwindElementCandidates,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    double minimum_edge_flux = std::numeric_limits<double>::max();
+    bool any_flux_negative = false;
+    const auto& element_boundary_geometry = this->GetGeometry().GenerateBoundariesEntities();
+
+    for (SizeType i = 0; i < rUpwindElementCandidates.size(); i++)
+    {
+        // get sorted node ids of neighbording elements
+        std::vector<size_t> neighbor_element_ids;
+        PotentialFlowUtilities::GetSortedIds<TDim, TNumNodes>(neighbor_element_ids, rUpwindElementCandidates[i].GetGeometry());
+
+        // find element which shares the upwind element nodes with current element
+        // but is not the current element
+        // extra check if its active
+
+        if (rUpwindElementCandidates[i].Id()!=this->Id()) {
+            for (auto& r_geom_boundary : element_boundary_geometry) {
+                std::vector<size_t> this_sorted_ids;
+                PotentialFlowUtilities::GetSortedIds<TDim, TNumNodes>(this_sorted_ids, r_geom_boundary);
+
+
+                if(std::includes(neighbor_element_ids.begin(), neighbor_element_ids.end(),
+                this_sorted_ids.begin(), this_sorted_ids.end()))
+                {
+                    const auto edge_normal = BaseType::GetEdgeNormal(r_geom_boundary);
+
+                    const double edge_flux = inner_prod(edge_normal, free_stream_velocity);
+
+                    if (edge_flux < 0) {
+                        any_flux_negative = true;
+                    }
+
+                    if(edge_flux < minimum_edge_flux && rUpwindElementCandidates[i].Is(ACTIVE)) {
+                        minimum_edge_flux = edge_flux;
+                        this->pSetUpwindElement(rUpwindElementCandidates(i));
+                    }
+                }
+            }
+        }
+    }
+
+    // If no upwind element is found, the element is an INLET element and the
+    // upwind element pointer points to itself
+    if (this->CheckUpwindElement() || !any_flux_negative)
```

Comment:
Sorry for dismissing the approval, but I noticed that some elements on the airfoil were detected as "INLET". Since the search of UpwindElements is only done on the ACTIVE elements, there might be some configurations were all the edge_fluxes of the active neighbours are positive:

![image](https://user-images.githubusercontent.com/32136457/156159085-9d809ef0-92c3-466f-a27f-908745219873.png)


So the element never enters to "SetUpwind" and it is detected as INLET.


In order to avoid this, I initilize the "minimum_flux" to numeric_limits::max(), and set as INLET those elements that dont have any surrounding element (active or non active) with negative flux.